### PR TITLE
Security - rc/init scripts

### PR DESCRIPTION
**Pull request by kmcdonell**
A bunch of changes to provide improved security, especially for PCP's rc/init/systemd scripts
- cleaner separation of parts run as user "root" from parts run as user "pcp"
- when running as user "root" do not modify (mkdir, cp, chmod, rm, ...) objects owned by the user "pcp"
- symmetry of logic for non-root scripts when run with systemd or init or any-other-replacement

### CODE REVIEW

**Review by natoscott — COMMENTED** [review]
Hi @kmcdonell - its looking good.  I'll add a few small nits inline shortly, but one higher level topic is I think the tmpfiles.d file needs to live elsewhere.  AIUI the /etc location is for sysadmins, and as packagers we are supposed to use the /usr/lib location ... (PCP_SYSTEMDTMPFILES_DIR)

```
$ rpm -ql pcp |grep tmpf
/etc/tmpfiles.d/pcp.conf
/usr/lib/tmpfiles.d/pcp.conf
```

The first is new of course, the latter we generate during the RPM builds (for rpm-ostree).  It'd be ideal if we can wrangle this content to be all in the second file.
I guess that means we need to shift this to be generated in the rpm/deb packaging rather than at this top level.

File: src/pmie/rc_pmie [hunk 1]
```diff
@@ -75,19 +51,6 @@ else
...
-    if [ -n "$PCP_DIR" ]
-    then
-	: running in a non-default installation, do not need to be root
-    else
-	echo "$prog:"'
-Error: You must be root (uid 0) to start or stop pmie.'
-	exit
-    fi
-fi
-
```
Comment (natoscott) [inline src/pmie/rc_pmie:90]: As discussed on slack, the restorecon invocation later in this file can go.

File: man/man1/runaspcp.1 [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from the
+.B \-s
+(or
+.BR \-\-shell)
+command line argument.
+.PP
+Normally
+.B runaspcp
+is used from the PCP ``init'' scripts that are launched
+as ``root'' but need to downgrade their priveleges when running
```
Comment (natoscott) [inline man/man1/runaspcp.1:48]: typo - "privileges"

File: man/man3i/__pmcleanmapdir.3 [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+will remove them:
+.IP + 2n
+the name should be numeric, else it should be equal
+.I special
+(if
+.I special
+is not NULL)
+.IP + 2n
+for a numeric directory entry it should be a regular file and
+and the name should be the PID of a running process
```
Comment (natoscott) [inline man/man3i/__pmcleanmapdir.3:61]: we have "and and" here (across two lines).

File: src/libpcp/src/util.c [hunk 4]
```diff
@@ -3162,3 +3166,235 @@ __pmLogFeaturesStr(__uint32_t features)
...
+{
+    switch (st_mode & S_IFMT) {
+	case S_IFBLK:	return "block device"; break;
+	case S_IFCHR:	return "character device"; break;
+	case S_IFDIR:	return "directory"; break;
+	case S_IFIFO:	return "FIFO/pipe"; break;
+	case S_IFLNK:	return "symlink"; break;
+	case S_IFREG:	return "regular file"; break;
+	case S_IFSOCK:	return "socket"; break;
+	default:	return "unknown?"; break;
```
Comment (natoscott) [inline src/libpcp/src/util.c:3182]: I think most of the "break;" use here (after return) can be dropped?  Personal style, I'd just keep the default: case as a "break;" and return unknown on the final line of the function.

File: src/pcp-reboot-init/pcp-reboot-init.sh [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# for more details.
+#
+# Intended as "run after reboot" to make sure transient things that PCP
+# needs have been setup ...
+#
+
+tmp=/var/tmp/pcp-reboot-init-$$
```
Comment (natoscott) [inline src/pcp-reboot-init/pcp-reboot-init.sh:19]: Should this use mktemp here like most PCP scripts do?  Think so.

**Comment by kmcdonell** [comment]
@natoscott re. tmpfiles.d ... we can just place the /run/pcp one in /usr/lib/tmpfiles.d/pcp-run-pcp.conf (avoids name conflict with rpm-generated ones) ... since /usr/lib/tmpfiles.d can exist without systemd, I'd prefer to use the explicit pathname rather than $(PCP_SYSTEMDTMPFILES_DIR) (which may translate to badness on a non-systemd platform).
Also I made it D rather then d in the .conf file so it goes away with package removal.

**Comment by natoscott** [comment]
> @natoscott re. tmpfiles.d ... we can just place the /run/pcp one in /usr/lib/tmpfiles.d/pcp-run-pcp.conf (avoids name conflict with rpm-generated ones) ... since /usr/lib/tmpfiles.d can exist without systemd, I'd prefer to use the explicit pathname rather than $(PCP_SYSTEMDTMPFILES_DIR) (which may translate to badness on a non-systemd platform). Also I made it D rather then d in the .conf file so it goes away with package removal.

@kenj yep, sounds good.  Another option might be to share a name with the pcp-reboot-init service?

**Comment by kmcdonell** [comment]
@natoscott OK, since the tmpfiles.d conf and the pcp-reboot-init script are effectively alternate ways of getting the same job done, I'm happy to rename the conf file.  This part is done in commits  1f28451aae and 97f85835.

File: src/pmie/rc_pmie (hunk 1 above)
Comment (kmcdonell) [reply src/pmie/rc_pmie:90]: _reboot_setup() all removed in commit 283b560b1

File: man/man3i/__pmcleanmapdir.3 (hunk 3 above)
Comment (kmcdonell) [reply man/man3i/__pmcleanmapdir.3:61]: Fixed in commit 7531b6404

File: man/man1/runaspcp.1 (hunk 2 above)
Comment (kmcdonell) [reply man/man1/runaspcp.1:48]: Fixed in commit 7531b6404

File: src/libpcp/src/util.c (hunk 4 above)
Comment (kmcdonell) [reply src/libpcp/src/util.c:3182]: Fixed in commit f29a6af00

File: src/pcp-reboot-init/pcp-reboot-init.sh (hunk 5 above)
Comment (kmcdonell) [reply src/pcp-reboot-init/pcp-reboot-init.sh:19]: Yep, fixed in commit 35600595c

**Pull request merged**